### PR TITLE
Support context, tensor, and pipeline parallelism with MinimalAsyncEP

### PR DESCRIPTION
**Pull request by wwwjn**
## Summary

Enable MinimalAsyncEP to compose with context parallelism, tensor parallelism, and pipeline parallelism.

This PR does not add backend-specific CP, TP, or PP dispatch algorithms. The required behavior comes from the common MoE and mesh contracts:

- #3996 makes `RoutedExperts` operate on a local sequence shard and return that local layout.
- #4080 physically pads the sequence before routing, so every rank dispatches the same number of materialized token rows.
- #3970 centralizes lifetime communication-buffer capacity inference for persistent EP backends.
- The sparse mesh places PP outside EP, so every pipeline stage owns an independent EP process group.

MinimalAsyncEP therefore continues to dispatch the local `x_TD` rows over its EP group and returns a combined tensor with the same local token shape. The common MoE sharding path handles CP and TP, while the pipeline runtime invokes the stage-local MinimalAsyncEP instances.

## Buffer capacity

MinimalAsyncEP now uses the shared lifetime capacity derivation:

```python
num_token_shards = context_parallel_degree * tensor_parallel_degree
if seq_len % num_token_shards != 0:
    raise ValueError(...)
num_max_tokens_per_rank = local_batch_size * seq_len // num_token_shards
```

CP and TP reduce the local token capacity because they shard the sequence dimension before MoE. The shared configuration requires `seq_len` to divide evenly across `CP * TP`; users must pad the configured sequence length when it does not. PP does not enter this calculation because it partitions model layers, not the token axis.

This replaces the backend-local `local_batch_size * seq_len` calculation, which overallocated once CP or TP was enabled. The shared setup also fills `hidden_dim`, preserves a user-specified capacity when it is large enough, and rejects an undersized capacity. MinimalAsyncEP keeps only its backend-specific dtype setup.

## Changes

- Remove the MinimalAsyncEP configuration restrictions for CP, TP, and PP.
- Include MinimalAsyncEP in shared EP buffer-capacity configuration.
- Add an H100 integration configuration for `DP-shard=2, CP=2, TP=2, EP=8`.
- Keep the test on `spmd_types` until the H100 suite migrates from the default DTensor backend.

### Tensor parallelism

Four H100s, 10 training steps:

```text
DP-shard=2, TP=2, EP=4
spmd_backend=spmd_types
compile disabled
full activation checkpointing
```

Loss decreased from `8.06603` to `3.86079`; final `grad_norm` was `1.6618`.

### Context parallelism

Four H100s, 10 training steps:

```text
DP-shard=2, CP=2, EP=4
spmd_backend=spmd_types
compile disabled
full activation checkpointing
```

Loss decreased from `8.00828` to `3.87139`; final `grad_norm` was `1.5916`.

### Pipeline parallelism

Four H100s with PyTorch `2.14.0.dev20260811+cu130`, two optimizer steps:

```text
DP-shard=2, PP=2, EP=2
1F1B with 8 microbatches
spmd_backend=spmd_types
compile disabled
full activation checkpointing
```

The run completed MinimalAsyncEP dispatch/combine, pipeline forward/backward, and PP-wide gradient-norm reduction. Loss decreased from `8.32359` to `6.38615`; `grad_norm` changed from `3.6094` to `4.4162`.

### Numerical comparison against AllToAll

The standard AllToAll dispatcher and MinimalAsyncEP were run for 10 steps with the same `DP-shard=2, PP=2, EP=2` topology, fused expert kernels, data, seed 42, and deterministic mode. Full-precision TensorBoard metrics show a maximum relative loss difference of `4.55e-4` and a maximum relative grad-norm difference of `2.60e-3`.

MinimalAsyncEP is not expected to be bitwise identical to AllToAll: AllToAll rounds each score-weighted expert row to BF16 and then performs deterministic scatter-add in expert order, while MinimalAsyncEP accumulates the top-k rows in FP32 slot order and rounds once.

Two diagnostics verify that this difference is numerical rather than token dropping:

- A temporary assertion reduced the active receive-row count across every stage-local EP group and required it to equal `EP_size * local_tokens * top_k`. It passed every forward and activation-recompute dispatch across all 10 steps.
- Replacing only MinimalAsyncEP's top-k combine arithmetic with the AllToAll arithmetic produced the exact same step-1 loss: `8.1351776123046875` for both backends. Native MinimalAsyncEP produced `8.1351785659790039`.

The exact eight-GPU `DP-shard=2, CP=2, TP=2, EP=8` CI configuration was not run locally because the host has seven available H100s. CP and TP were validated independently as listed above.

## Limitations

- MinimalAsyncEP requires the `spmd_types` backend. The default DTensor path currently hits the same `Shard(1) -> Partial(sum)` limitation tracked by #4110.
- MinimalAsyncEP requires full recomputation: full activation checkpointing for eager training or full memory policy for GraphTrainer.

### CODE REVIEW

File: torchtitan/models/common/token_dispatcher.py [hunk 1]
```diff
@@ -1177,8 +1177,8 @@ def combine(
 class MinimalAsyncEPTokenDispatcher(BaseEPTokenDispatcher):
     """Token dispatcher using MinimalAsyncEP for constrained EP communication.
 
-    This first integration supports EP with ``sp_size == 1`` only. TP/SP, CP,
-    PP, and padding are intentionally out of scope.
+    TP token sharding is supported independently of the dense sequence-parallel
+    setting. CP and PP are intentionally out of scope.
```
Comment (tianyu-l) [inline torchtitan/models/common/token_dispatcher.py]: I don't think we need to do anything for CP or PP?

**Review by tianyu-l — APPROVED** [review]
SGTM. Please wait for @xmfan 's approval.

Also cc @acisseJZhong I think the code related to padding will be gone once we move it to vllm wrapper.

File: torchtitan/models/common/token_dispatcher.py [hunk 2]
```diff
@@ -1360,6 +1365,23 @@ def combine(
             state.num_tokens,
             state.top_k,
         )
+        if self.sp_size > 1:
```
Comment (acisseJZhong) [inline torchtitan/models/common/token_dispatcher.py]: I am removing this in https://github.com/pytorch/torchtitan/pull/3996/changes

**Review by xmfan — APPROVED** [review]

File: torchtitan/distributed/minimal_async_ep/api.py [hunk 3]
```diff
@@ -151,10 +147,6 @@ def maybe_update_minimal_async_ep_config(model_config: Any, config: Any) -> None
         )
 
     for token_dispatcher_cfg in dispatcher_cfgs:
-        token_dispatcher_cfg.hidden_dim = model_config.dim
-        token_dispatcher_cfg.num_max_tokens_per_rank = (
```
Comment (acisseJZhong) [inline torchtitan/distributed/minimal_async_ep/api.py:155]: curious why this becomes https://github.com/pytorch/torchtitan/blame/4a93ee4e4bd72a71b1dae7d9555c764ffc5e18db/torchtitan/models/common/token_dispatcher.py#L1223 now?

Comment (wwwjn) [reply torchtitan/distributed/minimal_async_ep/api.py:155]: We size the buffer using life-time max number of local tokens, to avoid token drop. 

Before this change, MinimalAsyncEP only supports TP=1 CP=1, so in pre-training:
- Life-time max local tokens = (local_batch_size * seq_length) / (TP * CP) = local_batch_size * seq_length

After current change:
- Life-time max local tokens = (local_batch_size * seq_length) / (TP * CP) = local_batch_size * seq_length / TP

Comment (acisseJZhong) [reply torchtitan/distributed/minimal_async_ep/api.py:155]: why do we need 
```
        required_num_max_tokens_per_rank = training.local_batch_size * (
            (training.seq_len + num_token_shards - 1) // num_token_shards
        )
```
the ceil devision? I think we already made sure the token seq len is evenly divided by TP through padding. For CP I think we didn't pad now, can we just hardcode CP=1 for now?

File: torchtitan/models/common/token_dispatcher.py (hunk 1 above)
Comment (wwwjn) [reply torchtitan/models/common/token_dispatcher.py]: Agreed: no CP- or PP-specific dispatch algorithm is needed. The patch now only removes the old configuration guards and documents the existing composition. CP/TP reach MinimalAsyncEP as local sequence shards through the common MoE path; PP gives each pipeline stage its own EP process group. I validated CP with DP-shard=2, CP=2, EP=4 for 10 steps and PP with DP-shard=2, PP=2, EP=2 for two optimizer steps.

File: torchtitan/distributed/minimal_async_ep/api.py (hunk 3 above)
Comment (wwwjn) [reply torchtitan/distributed/minimal_async_ep/api.py:155]: CP is now validated, so I did not hardcode CP=1. This PR reuses the shared persistent-buffer bound rather than adding a MinimalAsyncEP-specific formula. CP and TP shard the sequence before MoE, so the maximum local rows are `local_batch_size * ceil_div(seq_len, CP * TP)`. For the currently divisible configurations, `ceil_div` is identical to exact division; it also remains a valid upper bound if the sharding path yields an uneven local shard. The CP=2 run completed 10 steps, and the routed-row conservation check passed on every dispatch in the separate 10-step numerical run.

File: torchtitan/models/common/token_dispatcher.py (hunk 2 above)
Comment (wwwjn) [reply torchtitan/models/common/token_dispatcher.py]: Rebased after #3996 landed. The old dispatcher-specific reconstruction is gone; this patch uses the common RoutedExperts local-shard path.

File: torchtitan/distributed/minimal_async_ep/api.py (hunk 3 above)
Comment (acisseJZhong) [reply torchtitan/distributed/minimal_async_ep/api.py:155]: IMO we should not do ceil devision if we already pad to even, we should raise explicit error if it's still uneven, and user should be responsible to pad even.

Comment (wwwjn) [reply torchtitan/distributed/minimal_async_ep/api.py:155]: You are right, we should check then error if uneven, given we already done padding in #4080, let me update

File: torchtitan/models/common/token_dispatcher.py (hunk 1 above)
Comment (wwwjn) [reply torchtitan/models/common/token_dispatcher.py]: Are you saying we should get CP and PP for free? I think CP and PP should work out of pocket, let me test and enable CP as well.
